### PR TITLE
fix(event_handler): security scheme unhashable list when working with router

### LINKED ISSUE #4375: Security configuration for route does not work for APIGatewayHttpRouter
### Expected Behaviour

Method `get_openapi_json_schema` from `APIGatewayHttpResolver` should work for routes from `APIGatewayHttpRouter` with the `security` parameter.

### Current Behaviour

When the `security` parameter is a part of the router (`APIGatewayHttpRouter`), it doesn't work - `/openapispec`  has: `{"message": "Internal Server Error"}`.

If I remove the `security` parameter from the router file, it works, and it means that the security configuration is correct only for `APIGatewayHttpResolver`.

Error in CloudWatch after `/openapispec ` execution:

`[ERROR] TypeError: unhashable type: 'list'
Traceback (most recent call last):
  File "/var/lang/lib/python3.12/importlib/__init__.py", line 90, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 995, in exec_module
  File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
  File "/var/task/app.py", line 14, in <module>
    from project_member_v1 import router as project_member_router
  File "/var/task/project_member_v1/router.py", line 27, in <module>
    @router.get(
  File "/opt/python/aws_lambda_powertools/event_handler/api_gateway.py", line 2412, in register_route
    self._routes_with_middleware[route_key] = []`

### Code snippet

```python
`
app = APIGatewayHttpResolver(enable_validation=True)
logger = Logger()
metrics = Metrics(namespace="Powertools")

OpenApiConfiguration.configure_swagger(app)

app.include_router(test_router.router, prefix="/v1/test")

@app.get("/hello", tags=["Test"], summary="An endpoint to test", security=[{"TestAuth": []}],)
def hello():
    return {"message": f"Hello!"}

@app.get("/openapispec", tags=["Docs"], summary="Get the OpenAPI specification")
def get_openapi():
    return app.get_openapi_json_schema(**OpenApiConfiguration.get_meta_data())


@logger.inject_lambda_context(correlation_id_path=correlation_paths.API_GATEWAY_REST)
@metrics.log_metrics(capture_cold_start_metric=True)
def lambda_handler(event: dict, context: LambdaContext) -> dict:
    return app.resolve(event, context)
`

Router file:
`
router = APIGatewayHttpRouter()

@router.get(
    "/<test_id>",
    tags=["Test"],
    summary="Test endpoint",
    security=[{"TestAuth": []}],
)
def get_test_item(test_id: str) -> TestItemResponse:
    return TestItemResponse(test_id= test_id)
`

`
class OpenApiConfiguration:
    @staticmethod
    def get_meta_data():
        return {
            "title": "TestAPI",
            "version": "0.0.1",
            Contact(name="Test", url=Url("test.com")),
            "description": "TestAPI",
            "terms_of_service": None,
            "license_info": None,
            "openapi_version": "3.0.1",
            "security_schemes": {
                "TestAuth": HTTPBearer(scheme="bearer", bearerFormat="JWT", description="JWT token authentication"),
            },
        }

    @staticmethod
    def configure_swagger(resolver: ApiGatewayResolver) -> None:
        resolver.enable_swagger(
            **OpenApiConfiguration.get_meta_data(),
            path="/swagger",
            swagger_base_url="https://cdnjs.cloudflare.com/ajax/libs/swagger-ui/5.11.0"
        )
`

Route `openspec` should returns a definition of 3 endpoints:
- `/hello` with `TestAuth` security
- `/openapispec` without security
- `/test/< test_id >`  with `TestAuth` security
```

### Possible Solution

_No response_

### Steps to Reproduce

1. Create instance of `APIGatewayHttpResolver` 
2. Create instance of `APIGatewayHttpRouter`
3. Implement `OpenApiConfiguration` and define security_schemes
4. Add route to APIGatewayHttpRouter with security 
5. Add route to openapispec: 
`@app.get("/openapispec", tags=["Docs"], summary="Get the OpenAPI specification")
def get_openapi():
    return app.get_openapi_json_schema(**OpenApiConfiguration.get_meta_data())`
6. Execute `/openapispec`

### Powertools for AWS Lambda (Python) version

latest

### AWS Lambda function runtime

3.12

### Packaging format used

Lambda Layers

### Debugging logs

_No response_

### LINKED ISSUE #4460: Bug: Security schemes are not validated when using operation level
### Expected Behaviour

The OpenAPI specification expects `security schemes` to define what type of authentication that API will support and then customers can define the operations that will use that authentication or even whether all operations will use the same authentication.

When using this code, the OpenAPI utility should validate the security defined in operation level and throw an exception as it is a schema validation error. But this is not happening.

```python
from aws_lambda_powertools.event_handler import APIGatewayRestResolver

app = APIGatewayRestResolver(enable_validation=True)

@app.get("/hello", security=[{"apiKey": []}])
def get_schema():
    return app.get_openapi_json_schema()

def lambda_handler(event, context):
    return app.resolve(event, context)
```

### Current Behaviour

This is currently not being validated and customers receive an error when trying to use the OpenAPI specification.

### openapischema

```json
{
   "openapi":"3.0.3",
   "info":{
      "title":"Powertools API",
      "version":"1.0.0"
   },
   "servers":[
      {
         "url":"/"
      }
   ],
   "paths":{
      "/hello":{
         "get":{
            "summary":"GET /hello",
            "operationId":"get_schema_hello_get",
            "responses":{
               "200":{
                  "description":"Successful Response",
                  "content":{
                     "application/json":{
                        
                     }
                  }
               },
               "422":{
                  "description":"Validation Error",
                  "content":{
                     "application/json":{
                        "schema":{
                           "$ref":"#/components/schemas/HTTPValidationError"
                        }
                     }
                  }
               }
            },
            "security":[
               {
                  "apiKey":[
                     
                  ]
               }
            ]
         }
      }
   },
   "components":{
      "schemas":{
         "HTTPValidationError":{
            "properties":{
               "detail":{
                  "items":{
                     "$ref":"#/components/schemas/ValidationError"
                  },
                  "type":"array",
                  "title":"Detail"
               }
            },
            "type":"object",
            "title":"HTTPValidationError"
         },
         "ValidationError":{
            "properties":{
               "loc":{
                  "items":{
                     "anyOf":[
                        {
                           "type":"string"
                        },
                        {
                           "type":"integer"
                        }
                     ]
                  },
                  "type":"array",
                  "title":"Location"
               },
               "type":{
                  "type":"string",
                  "title":"Error Type"
               }
            },
            "type":"object",
            "required":[
               "loc",
               "msg",
               "type"
            ],
            "title":"ValidationError"
         }
      }
   }
}
```

Error:

image

### Code snippet

```python
from aws_lambda_powertools.event_handler import APIGatewayRestResolver

app = APIGatewayRestResolver(enable_validation=True)

@app.get("/hello", security=[{"apiKey": []}])
def get_schema():
    return app.get_openapi_json_schema()

def lambda_handler(event, context):
    return app.resolve(event, context)
```

### Possible Solution

_No response_

### Steps to Reproduce

Use this code

```python
from aws_lambda_powertools.event_handler import APIGatewayRestResolver

app = APIGatewayRestResolver(enable_validation=True)

@app.get("/hello", security=[{"apiKey": []}])
def get_schema():
    return app.get_openapi_json_schema()

def lambda_handler(event, context):
    return app.resolve(event, context)
```

### Powertools for AWS Lambda (Python) version

latest

### AWS Lambda function runtime

3.11

### Packaging format used

PyPi

### Debugging logs

_No response_

### PR DESCRIPTION
**Pull request by leandrodamascena**
**Issue number:** #4375 

### Changes

This pull request ensures that the `get_openapi_json_schema` method can correctly generate the OpenAPI JSON schema for routes that include security parameters when using the `APIGatewayHttpRouter`.

Previously, when the security parameter was included in a router, any endpoint would return an "Internal Server Error".

```sh
[ERROR] TypeError: unhashable type: 'list' Traceback (most recent call last): File "/var/lang/lib/python3.12/importlib/__init__.py", line 90, in import_module return _bootstrap._gcd_import(name[level:], package, level) File "<frozen importlib._bootstrap>", line 1387, in _gcd_import File "<frozen importlib._bootstrap>", line 1360, in _find_and_load File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked File "<frozen importlib._bootstrap>", line 935, in _load_unlocked File "<frozen importlib._bootstrap_external>", line 995, in exec_module File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed File "/var/task/app.py", line 14, in <module> from project_member_v1 import router as project_member_router File "/var/task/project_member_v1/router.py", line 27, in <module> @router.get( File "/opt/python/aws_lambda_powertools/event_handler/api_gateway.py", line 2412, in register_route self._routes_with_middleware[route_key] = []
```

#### app.py
```python
from aws_lambda_powertools.event_handler import APIGatewayHttpResolver
from router import OpenApiConfiguration, router


app = APIGatewayHttpResolver(enable_validation=True)
OpenApiConfiguration.configure_swagger(app)


app.include_router(router, prefix="/v1/test")

@app.get("/openapispec", tags=["Docs"], summary="Get the OpenAPI specification", security=[{"TestAuth": []}])
def get_openapi():
    return app.get_openapi_json_schema(**OpenApiConfiguration.get_meta_data())

def lambda_handler(event, context):
    return app.resolve(event, context)
```

#### router.py
```python
from aws_lambda_powertools.event_handler import APIGatewayHttpResolver
from aws_lambda_powertools.event_handler.api_gateway import Router
from aws_lambda_powertools.event_handler.openapi.models import HTTPBearer

router = Router()

@router.get(
    "/<test_id>",
    tags=["Test"],
    summary="Test endpoint",
    security=[{"TestAuth": ["a"]}],
)
def get_test_item(test_id: str) :
    return "OK"

class OpenApiConfiguration:
    @staticmethod
    def get_meta_data():
        return {
            "title": "TestAPI",
            "version": "0.0.1",
            "description": "TestAPI",
            "terms_of_service": None,
            "license_info": None,
            "openapi_version": "3.0.1",
            "security_schemes": {
                "TestAuth": HTTPBearer(scheme="bearer", bearerFormat="JWT", description="JWT token authentication"),
            },
        }

    @staticmethod
    def configure_swagger(resolver: APIGatewayHttpResolver) -> None:
        resolver.enable_swagger(
            **OpenApiConfiguration.get_meta_data(),
            path="/swagger",
            swagger_base_url="https://cdnjs.cloudflare.com/ajax/libs/swagger-ui/5.11.0"
        )
```

### User experience

> Please share what the user experience looks like before and after this change

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda/python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

File: aws_lambda_powertools/event_handler/util.py
Comment (heitorlessa) [inline aws_lambda_powertools/event_handler/util.py]: break this into multiple lines to make it easier to read plz

File: tests/functional/event_handler/test_openapi_security.py
Comment (heitorlessa) [inline tests/functional/event_handler/test_openapi_security.py]:
```suggestion
    top_level_security = schema.security
    path_level_security = schema.paths["/"].get.security

    assert top_level_security is None
    assert path_level_security[0] == {"apiKey": []}
```

* tip: careful with shadowing variables with the same name. 
* tip: when you have multiple assertion, often times you might be able to narrow down to one or two, otherwise split tests. For example, if you're asserting operation level security scheme for a given value, you are already implicitly asserting that `is not None`, `it has at least 1 item` - you can therefore keep it down to a single assert.

File: tests/functional/event_handler/test_openapi_security.py
Comment (heitorlessa) [inline tests/functional/event_handler/test_openapi_security.py]: is that necessary for the test?

File: aws_lambda_powertools/event_handler/api_gateway.py
Comment (heitorlessa) [inline aws_lambda_powertools/event_handler/api_gateway.py:2405]: do we anticipate another non-hashable parameter we need to resolve now too?

I remember `tags` was one of them, now `security`. I dunno the implementation yet so I wonder if we need to stop and build a more robust solution OR if `security` is the last piece we overlooked

**Review by heitorlessa — CHANGES_REQUESTED** [review]
one tiny change, a few suggestions to improve tests, and one major question to prevent another bug

File: aws_lambda_powertools/event_handler/util.py
Comment (leandrodamascena) [reply aws_lambda_powertools/event_handler/util.py]: I understand what you mean, but black is formatting this line, so it makes sense to follow the linter formatting. WDYT?

File: tests/functional/event_handler/test_openapi_security.py
Comment (leandrodamascena) [reply tests/functional/event_handler/test_openapi_security.py]: Yep, this is because we need to configure the route with the security scheme.

File: aws_lambda_powertools/event_handler/api_gateway.py
Comment (leandrodamascena) [reply aws_lambda_powertools/event_handler/api_gateway.py:2405]: At this moment, security is the last piece we overlooked. I can't say that we won't change this function or that we will never add another parameter, but at the moment we have the OpenAPI implementation almost complete, missing things like vendor extension and cookies, for example. But I don't think both will cause problems here.

File: tests/functional/event_handler/test_openapi_security.py
Comment (leandrodamascena) [reply tests/functional/event_handler/test_openapi_security.py]: Thanks a lot for this tips, @heitorlessa! I fixed this here and in another test.

File: aws_lambda_powertools/event_handler/util.py
Comment (heitorlessa) [reply aws_lambda_powertools/event_handler/util.py]: use multiple expressions in this case. I'm gonna pull the PR to send a suggestion as Web version is tricky

File: aws_lambda_powertools/event_handler/api_gateway.py
Comment (heitorlessa) [reply aws_lambda_powertools/event_handler/api_gateway.py:2405]: gotcha. Let's do this, if it happens with yet another parameter, we will stop and prioritize investigating a refactoring. The encoder is a good example of an area that we should've used a singledispatch to avoid a myriad of if/elif we overlooked at launch (issue with extremely large PRs).

**Comment by heitorlessa** [comment]
Adding a summary shortly after our call that led to important discoveries

File: tests/functional/event_handler/test_openapi_security.py
Comment (heitorlessa) [inline tests/functional/event_handler/test_openapi_security.py]: **change**: move these into a top-level variable like `top_level_security_schemes` to ease reading the intent (pre-setup), please.

Later it'll make it even clearer when we have a new API to configure OpenAPI.

File: tests/functional/event_handler/test_openapi_security.py
Comment (heitorlessa) [inline tests/functional/event_handler/test_openapi_security.py]: **change**: move these into a top-level variable like `top_level_security_schemes` to ease reading the intent (pre-setup), please.

Later it'll make it even clearer when we have a new API to configure OpenAPI.

File: tests/functional/event_handler/test_openapi_security.py
Comment (heitorlessa) [reply tests/functional/event_handler/test_openapi_security.py]: **change**: move these into a top-level variable like `top_level_security_schemes` to ease reading the intent (pre-setup), please.

Later it'll make it even clearer when we have a new API to configure OpenAPI.

File: aws_lambda_powertools/event_handler/util.py
Comment (heitorlessa) [reply aws_lambda_powertools/event_handler/util.py]: Best I could tinker right now - got rid of the extra wrapping as it was unnecessary. Tests passing but lemme know otherwise.

I broke by:

1. Group all items that are hashable
2. We're interested in immutable values so only use `frozenset` in the value itself
3. Create a `frozenset` out of all hashable items
4. Hash it

That way we also don't need to use `_FrozenDict` to wrap some of it too.

---

```suggestion
        hashable_items = []
        for item in self:
            hashable_items.extend((key, frozenset(value)) for key, value in item.items())
        
        return hash(frozenset(hashable_items))
```

**Review by heitorlessa — CHANGES_REQUESTED** [review]
last changes before merging the PR.

* Suggested a multi-expression and slightly better perf for hashing (along w/ explanation)
* Suggested a change in tests to make it more explicit that there is a `setup` for OpenAPI until we have a new API to solve the ambiguity
* We're missing a test for Security Scheme mismatch. Top-level says `ApiKey` but path level uses a `JWT`.

If you've got time, it'd be great to already address other pieces we missed prior to this PR

* An OpenAPI Schema specific exception, not `ValueError`. The exceptions we're returning can be more actionable too (we can start with the security scheme)

Thank you so much for the hard work on this <3

**Comment by heitorlessa** [comment]
For anyone reading this later...

As part of the security scheme bug fix review (#4421), we found that the responsibilities between OpenAPI and SwaggerUI are ambiguous, leading to confusion among customers and maintainers _at times_.

As of now, these are common use cases:

1. **Setting up OpenAPI**. Customers use a combination of implicit type annotations, OpenAPI params at the route level, and extra params via `enable_swagger`.
2. **Setting up SwaggerUI**. Customers use `enable_swagger` to render the implicit  OpenAPI Schema in a web app _(SwaggerUI)_, customize assets, etc.
3. **Serializing OpenAPI JSON Schema**. Customers create an explicit route _e.g.,_ `/schema`, and use [`app.get_openapi_json_schema`](https://docs.powertools.aws.dev/lambda/python/latest/core/event_handler/api_gateway/#global-openapi-security-schemes) to serialize it. Those using SwaggerUI can download it at the homepage _(https://endpoint.url/swagger?format=json)_.

We're confusing customers on 1 and 2, because `enable_swagger` or `get_openapi_*schema` is the only way to further extend / override OpenAPI. Customers **should not** need to use `enable_swagger` or `get_openapi_*schema` to further customize OpenAPI.

**How do we fix it in a backwards compatible way?**

> need two new issues

* Introduce a new API to configure / extend OpenAPI in the resolver instance _(`app`)_
* Improve docs by surfacing that API instead of `get_openapi_*schema` and `enable_swagger` - [example](https://docs.powertools.aws.dev/lambda/python/latest/core/event_handler/api_gateway/#customizing-openapi-metadata)
* Rename and revamp OpenAPI docs sections to match the use-case driven docs  we have. For now, it should be easier to find how to
  * [ ] extend OpenAPI spec
  * [ ] export OpenAPI schema programmatically
  * [ ] override OpenAPI schema when exporting
  * [ ] when to override OpenAPI metadata in SwaggerUI

File: aws_lambda_powertools/event_handler/util.py
Comment (leandrodamascena) [reply aws_lambda_powertools/event_handler/util.py]: good catch!

File: tests/functional/event_handler/test_openapi_security.py
Comment (leandrodamascena) [reply tests/functional/event_handler/test_openapi_security.py]: Done

File: tests/functional/event_handler/test_openapi_security.py
Comment (leandrodamascena) [reply tests/functional/event_handler/test_openapi_security.py]: Done

File: tests/functional/event_handler/test_openapi_security.py
Comment (leandrodamascena) [reply tests/functional/event_handler/test_openapi_security.py]: Done

**Review by leandrodamascena — COMMENTED** [review]
Thanks for the review and suggestions @heitorlessa!

**Comment by heitorlessa** [comment]
checking in 👀 ;)

File: aws_lambda_powertools/event_handler/api_gateway.py
Comment (leandrodamascena) [inline aws_lambda_powertools/event_handler/api_gateway.py:1654]: Hey @heitorlessa! I've updated the exception from `ValueError` to `SchemaValidationError` as it is a schema validation and we should use the appropriate exception name. I don't think that customers are using this, but we can revert in any case. WDYT?

Comment (heitorlessa) [reply aws_lambda_powertools/event_handler/api_gateway.py:1654]: great thanks! I'll look into the messaging (what we tell customers in the exception) after lunch

File: aws_lambda_powertools/event_handler/util.py
Comment (heitorlessa) [inline aws_lambda_powertools/event_handler/util.py]: plz add more details on the logic here, it's easy to miss the difference between the two by looking at the `all` logic.

File: aws_lambda_powertools/event_handler/util.py
Comment (heitorlessa) [inline aws_lambda_powertools/event_handler/util.py]:
```suggestion
def _validate_openapi_security_parameters(
```

make it private to avoid breakage, as we should move these type of validations into a Schema Validator (missing?) once we have one

File: aws_lambda_powertools/event_handler/util.py
Comment (heitorlessa) [inline aws_lambda_powertools/event_handler/util.py]:
```suggestion
    
    security_schemes = security_schemes or {}

    security_schema_match = all(
        key in security_schemes for sec in security for key in sec
    )

    return bool(security_schema_match and security_schemes)
```

A few changes:

* **Always guard against a `None`**. Otherwise the `all` will raise `TypeError: argument of type 'NoneType' is not iterable`
* **Break statements when there are two four loops**. It's easy to accidentally get into a [quadratic situation](https://accidentallyquadratic.tumblr.com/). The exercise of breaking it allows you to have another look at the logic to see if there's another way to do it w/o a double loop (not in this case). At worst, it improves readability
* **Use intentional var names on validation**. For schema validation logic, it's easy to get lost when coming back to code as most of them look the same _(e.g., iterating through a data structure)_. Breaking statements, using var names for conditions, or breaking into multiple functions based on conditions make it easier to maintain later.

Q: is there a better naming we can use to make it less ambiguous for these function parameters?

File: aws_lambda_powertools/event_handler/util.py
Comment (heitorlessa) [inline aws_lambda_powertools/event_handler/util.py]:
```suggestion
        Whether list of security schemes match allowed security_schemes.
```

is that what you meant?

File: aws_lambda_powertools/event_handler/api_gateway.py
Comment (heitorlessa) [inline aws_lambda_powertools/event_handler/api_gateway.py]: QQ: As a rule of thumb, if `route.security` is None we don't want to proceed with any validation? 

If the answer is yes, I'd suggest encapsulating that as part of the validator function saving an extra `if` since `and` is an extra if..

File: aws_lambda_powertools/event_handler/api_gateway.py
Comment (heitorlessa) [inline aws_lambda_powertools/event_handler/api_gateway.py]: anything else we think it'd be useful to tell them? e.g., anything to look to fix it in the route X?

If not, leave as is.

File: tests/functional/event_handler/test_openapi_security.py
Comment (heitorlessa) [inline tests/functional/event_handler/test_openapi_security.py:43]: Niiiice!!

File: tests/functional/event_handler/test_openapi_security.py
Comment (heitorlessa) [inline tests/functional/event_handler/test_openapi_security.py]:
```suggestion
    # AND top level security is defined as HTTPBearer
```

File: tests/functional/event_handler/test_openapi_security.py
Comment (heitorlessa) [inline tests/functional/event_handler/test_openapi_security.py]:
```suggestion
    # AND a route with a security scheme defined
```

File: tests/functional/event_handler/test_openapi_security.py
Comment (heitorlessa) [inline tests/functional/event_handler/test_openapi_security.py]:
```suggestion
    # AND a route with a security scheme using HTTPBearer
```

**Review by heitorlessa — CHANGES_REQUESTED** [review]
THANK YOU for working out the missing tests <3

Made last round of suggestions and another accidental bug in the validator. We can merge afterwards

File: aws_lambda_powertools/event_handler/util.py
Comment (leandrodamascena) [reply aws_lambda_powertools/event_handler/util.py]: >Always guard against a None. Otherwise the all will raise TypeError: argument of type 'NoneType' is not iterable

In this case will not because the first expression test if security_schemes is defined. But your code makes it more clear.

File: aws_lambda_powertools/event_handler/api_gateway.py
Comment (leandrodamascena) [reply aws_lambda_powertools/event_handler/api_gateway.py]: In this case we will have `if` here or there, right? So, maybe it doesn't make much difference because we will always have if to check whether the route is `None` or not. I think the code and the intention of what we want to do will be more explicit now. WDYT?

File: aws_lambda_powertools/event_handler/api_gateway.py
Comment (leandrodamascena) [reply aws_lambda_powertools/event_handler/api_gateway.py]: I think this message is very explicit. Maybe add a `See: https://docs.powertools.aws.dev/lambda/python/latest/core/event_handler/api_gateway/#security-schemes`?

File: aws_lambda_powertools/event_handler/util.py
Comment (leandrodamascena) [reply aws_lambda_powertools/event_handler/util.py]: Great catch!

File: aws_lambda_powertools/event_handler/api_gateway.py
Comment (heitorlessa) [reply aws_lambda_powertools/event_handler/api_gateway.py]: Yes, let's add that. The section itself deserves improvement too, as we can now mention that a mismatch will raise a validation error.

File: aws_lambda_powertools/event_handler/api_gateway.py
Comment (heitorlessa) [reply aws_lambda_powertools/event_handler/api_gateway.py]: > in this case we will have `if` here or there, right?

You're 10000000% right. I assumed Python would not short-circuit and waste resources evaluating both.. but [IT DOES](https://docs.python.org/3/library/stdtypes.html#boolean-operations-and-or-not).

I'm gonna resolve it.

My personal preference is to always encapsulate validation within the validator to avoid surprises, but changing or not here doesn't lead to any bug :)

File: aws_lambda_powertools/event_handler/util.py
Comment (heitorlessa) [inline aws_lambda_powertools/event_handler/util.py]:
```suggestion
    security_schemes: Optional[Dict[str, "SecurityScheme"]] = None,
```

If the type annotation is correct, you're missing a `None`

File: aws_lambda_powertools/event_handler/util.py
Comment (heitorlessa) [reply aws_lambda_powertools/event_handler/util.py]: DUh, of course, I missed the `and` again.

File: docs/core/event_handler/api_gateway.md
Comment (heitorlessa) [inline docs/core/event_handler/api_gateway.md]:
```suggestion

Security schemes are declared at the top-level first. You can reference them globally or on a per path _(operation)_ level. **However**, if you reference security schemes that are not defined at the top-level it will lead to a `SchemaValidationError` _(invalid OpenAPI spec)_.
```

**Review by heitorlessa — APPROVED** [review]

**Comment by heitorlessa** [comment]
suuuuperb working in going through bugs, finding new ones, and improving that section of the docs. It's also great that we identified the whole OpenAPI docs need a big revamp, **and** that we need a new API to make this easier and less confusing <3

who's awesome!?

**Pull request merged**
